src/Storage/Device/Local.php

```diff
@@ -69,46 +72,112 @@ public function getPath($filename): string
      *
      * @param string $source
      * @param string $path
+     * @param int $chunk
+     * @param int $chunks
+     * @param array $metadata
      *
      * @throws \Exception
      *
-     * @return bool
+     * @return int
      */
-    public function upload($source, $path): bool
+    public function upload(string $source, string $path, int $chunk = 1, int $chunks = 1, array &$metadata = []): int
     {
         if (!\file_exists(\dirname($path))) { // Checks if directory path to file exists
             if (!@\mkdir(\dirname($path), 0755, true)) {
                 throw new Exception('Can\'t create directory: ' . \dirname($path));
             }
         }
 
-        if (\move_uploaded_file($source, $path)) {
-            return true;
+        //move_uploaded_file() verifies the file is not tampered with
+        if($chunks === 1) {
+            if (!\move_uploaded_file($source, $path)) {
+                throw new Exception('Can\'t upload file ' . $path);
+            }
+            return $chunks;
         }
+        $tmp = \dirname($path) . "/tmp/chunks.log";
 
-        return false;
+        if (!\file_exists(\dirname($tmp))) { // Checks if directory path to file exists
+            if (!@\mkdir(\dirname($tmp), 0755, true)) {
+                throw new Exception('Can\'t create directory: ' . \dirname($tmp));
+            }
+        }
+        if(!file_put_contents($tmp, "{$chunk}\n", FILE_APPEND)) {
+            throw new Exception('Can\'t write chunk log ' . $tmp);
```

Comment:
What do we mean by a chunk log ? Are we referring to actual log messages ?

Reply:
Nope, we have a log file where we write when each chunk is received, which is later used to count total chunks received and create a complete file when all the chunks are received. If we fail to write to this file, we will not be able to create whole file when all chunks have been received.